runtime/src/commonMain/kotlin/pbandk/internal/binary/BinaryMessageMarshaller.kt

```diff
@@ -0,0 +1,120 @@
+package pbandk.internal.binary
+
+import pbandk.*
+import pbandk.wkt.*
+import kotlin.Any
+import kotlin.reflect.KProperty1
+
+internal open class BinaryMessageMarshaller(private val wireMarshaller: BinaryWireMarshaller) : MessageMarshaller {
+    override fun <T : Message> writeMessage(message: T) {
+        for (fd in message.fieldDescriptors) {
+            @Suppress("UNCHECKED_CAST")
+            val value = (fd.value as KProperty1<T, *>).get(message)
+
+            if (!fd.type.hasPresence && fd.type.isDefaultValue(value)) {
+                continue
+            }
+
+            value?.let { writeFieldValue(fd.number, fd.type, it) }
+        }
+
+        for (field in message.unknownFields.values) {
+            writeUnknownFieldValue(field.fieldNum, field.value)
+        }
+    }
+
+    private fun writeFieldValue(fieldNum: Int, type: FieldDescriptor.Type, value: Any) {
+        when (type) {
+            is FieldDescriptor.Type.Primitive<*> -> wireMarshaller.writePrimitiveValue(fieldNum, type, value)
+
+            is FieldDescriptor.Type.Message<*> -> when (type.messageCompanion) {
+                DoubleValue.Companion -> writeWrapperValue(fieldNum, type, value as Double, Sizer::doubleSize)
+                FloatValue.Companion -> writeWrapperValue(fieldNum, type, value as Float, Sizer::floatSize)
+                Int64Value.Companion -> writeWrapperValue(fieldNum, type, value as Long, Sizer::int64Size)
+                UInt64Value.Companion -> writeWrapperValue(fieldNum, type, value as Long, Sizer::uInt64Size)
+                Int32Value.Companion -> writeWrapperValue(fieldNum, type, value as Int, Sizer::int32Size)
+                UInt32Value.Companion -> writeWrapperValue(fieldNum, type, value as Int, Sizer::uInt32Size)
+                BoolValue.Companion -> writeWrapperValue(fieldNum, type, value as Boolean, Sizer::boolSize)
+                StringValue.Companion -> writeWrapperValue(fieldNum, type, value as String, Sizer::stringSize)
+                BytesValue.Companion -> writeWrapperValue(fieldNum, type, value as ByteArr, Sizer::bytesSize)
+                else -> writeMessageValue(fieldNum, value as Message)
+            }
+            is FieldDescriptor.Type.Enum<*> -> wireMarshaller.writeEnum(fieldNum, value as Message.Enum)
+
+            is FieldDescriptor.Type.Repeated<*> -> writeRepeatedValue(
+                fieldNum,
+                value as List<*>,
+                type.valueType,
+                type.packed
+            )
+
+            is FieldDescriptor.Type.Map<*, *> -> writeMapValue(fieldNum, value as Map<*, *>, type)
+        }
+    }
+
+    private fun <T : Any> writeWrapperValue(
+        fieldNum: Int,
+        type: FieldDescriptor.Type.Message<*>,
+        value: T,
+        sizeFn: (T) -> Int
+    ) {
+        val valueType = type.messageCompanion.fieldDescriptors[0].type
+        if (valueType.isDefaultValue(value)) {
```

Comment:
Theres another default value check up on line 14.  Is one of these not needed?

Reply:
They're slightly different for wrapper types. Using the `StringValue` wrapper type as an example:
* The call on line 14 is doing `!fd.type.hasPresence && fd.type.isDefaultValue(value)`. `StringValue` is a message type and message types always "have presence" (i.e. they're nullable), so the `fd.type.isDefaultValue(value)` check won't even be called on line 14.
* Here on line 62, we've looked up the underlying type of the `StringValue` wrapper: `String`. We're now checking if `value` is the default value of `String` (not `StringValue`). If it is, we'll omit writing the empty string on the wire. But we're still writing the `StringValue` message header so the receiver knows that the value was non-null.